### ISSUE TITLE
Video cut off from SVTPlay

### ISSUE THREAD

**Issue by R3C01L-SWE** [spaam/svtplay-dl#1365]
### svtplay-dl versions:

svtplay-dl 3.5 and 3.6 

### Operating system and Python version:

MacOS X Big Sur Python 2.7.16 Linux Mint Python 2.7.18

### What is the issue:

For the moment this will be a very vague error report.

But basically what I am doing is downloading the whole archive of this https://www.svtplay.se/melodifestivalen-oppet-arkiv

But it has also happened on other shows from SVT. Never from TV4Play or Viafree though. But TV4Play instead suffer from the constant stuttering problem that has already been reported quite some time ago... But that's a whole other story...

The problem I'm seeing right now is that the program seems to fetch both audio and video streams to 100% but still in the end I end up with a broken muxed file.

The video cuts of at random parts and the audio then continues playing until the end.

This doesn't happen every time but out of 13 files that I have fetched so far 4 are complete so it's quite a big problem.

I have tried it on 2 systems and both suffer from the same problem it seems.

I am currently running the program using the --verbose option to try and see if it spits out anythings special during the process.

So I will get back with more information later.

**Comment by R3C01L-SWE** [comment]
So after some more digging I don't think this is a svtplay-dl error after all.

It is more SVTPlay representing svtplay-dl with a broken download that it then downloads and happily muxed together. So I guess I will have to babysit and after check all SVTPlay downloads which is a bit of a PITA but oh well.

Repeating the download usually fixes it so my guess is it gets pulled from another CDN point that isn't carrying a broken file.

**Comment by R3C01L-SWE** [comment]
Here is another example from another show. Nothing seems wrong in the download itself and --verbose doesn't really show anything but the file is completely messed up.

![Tartan-6-File](https://user-images.githubusercontent.com/82805383/115787981-46666d00-a3c3-11eb-92dd-1f2cd69ac9e9.png)
![Tartan-6 Terminal](https://user-images.githubusercontent.com/82805383/115787986-46ff0380-a3c3-11eb-9dd1-f89fb98cadf0.png)

**Comment by R3C01L-SWE** [comment]
I guess we can close this one now. I got tired of trying to find where the error lies. But since the download seems to complete it has to be something in post processing / muxing that breaks it all. 

Anyhow sadly my solution was to switch to YouTube-dl and so far zero files have come out damaged. And that is after grabbing around 1TB of data and around 60 shows and programs..

**Comment by Sopor** [comment]
There have been a couple of errors when using older version of ffmpeg. What version do you use? You need to use ffmpeg 4 or higher.

**Comment by R3C01L-SWE** [comment]
@Sopor 

The Mint box uses 4.2.4
Mac uses 4.4 

So it shouldn't be that. 

One thing I did notice with the switch to YouTube-dl is that it is complaining a bit about errors in the audio file and fixes them on the fly.

Not sure if this could be what is causing the incomplete files with svtplay-dl though.

**Comment by towil1** [comment]
I have this problem with svtplay too, in some weeks it is worse than others.  Either the video or the audio stream gets truncated, so added a routine to print out a warning if there is a difference in the stream durations:

 `diff --git a/lib/svtplay_dl/postprocess/__init__.py b/lib/svtplay_dl/postprocess/__init__.py
index cc369ba4..73dc888f 100644
--- a/lib/svtplay_dl/postprocess/__init__.py
+++ b/lib/svtplay_dl/postprocess/__init__.py
@@ -1,8 +1,10 @@
 import logging
 import os
+import sys
 import platform
+import subprocess
 import re
-from json import dumps
+from json import dumps, loads
 from random import sample
 from re import match
 from shutil import which
@@ -13,6 +15,30 @@ from requests import Timeout
 from svtplay_dl.utils.output import formatname
 from svtplay_dl.utils.proc import run_program

+# Check if merged file is OK
+def check_durations(file):
+    logging.info("Checking merged file.")
+    cmd = ["ffprobe","-v","quiet", "-print_format","json","-show_streams","file:"+str(file)]
+    p = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, stdin=subprocess.PIPE)
+    stdout, stderr = p.communicate()
+    if p.returncode != 0:
+        stderr = stderr.decode("utf-8", "replace")
+        msg = stderr.strip().split("\n")[-1]
+        logging.error("Something went wrong: {0}".format(msg))
+        logging.error("Command: {0}".format(cmd))
+        return
+
+    jsondata = loads(stdout)
+    vduration=jsondata["streams"][0]["duration"]
+    aduration=jsondata["streams"][1]["duration"]
+    avdiff = float(vduration) - float(aduration)
+
+    if abs(avdiff) > 1.0:
+        sys.stdout.write("\033[1;31m") # Bright red
+        msg = "video duration "+vduration+" vs audio "+aduration+" difference " + str(avdiff)
+        logging.error("The merged file is corrupted: {0}".format(msg))
+        sys.stdout.write("\033[0;0m") # Reset colour
+        sys.exit(2)
`  

Additionally for svtplay.se I've added retries to the request method in http.py  if the recieved segement length is less than expected, which also seems to help:

`        # Retry if we did not received the expected length
        if 'svt-vod' in url and 'content-length' in res.headers:
            while len(res.content) < int(res.headers['content-length']):
                self.retries += 1
                sys.stderr.write(" "+str(self.retries)+" retries")
                res = Session.request(self, method, url, verify=self.verify, proxies=proxy, *args, **kwargs)
`

Hope you find this useful, and maybe @Sopor would want to incorporate a cleaned up version of this

**Comment by spaam** [comment]
@towil1  interesting,  so the size of the content we get is different from the conent-length  🤔

**Comment by towil1** [comment]
@spaam, yes this is at least part of the explanation for the issues with svtplay.se.  Most of the time it is OK, but some days I see multiple retries of the same stream.  Thought it was just me and maybe because I'm in the UK and use a proxy in Stockholm. (And sorry, I meant @spaam above).

**Comment by spaam** [comment]
ahh so it could be a proxy issue and not a svtplay-dl issue 🤔  
have you tried using --http-header "x-forwarded-for=Swedish-ip" (replace swedish-ip with a real one...) and see if that works better? 

@R3C01L-SWE do you also use a proxy?

i guess i should update the issue template so people dont forgot to mention if they are using proxy. :(

**Comment by towil1** [comment]
No, as I have a "transparent" SOCKS5 proxy (haproxy)  and then do "tsocks svtplay-dl".  It is more like a network glitch, and it always eventually works, either via these inline segment retries, or by downloading the whole episode again.

**Comment by R3C01L-SWE** [comment]
@spaam 

No proxy involved. Both boxes are sitting on Swedish 1GE links provided by A3 and Telia. 

One box sits at my summer place and the other the Mac is with me at my current place of living.
